packages/lambda/src/launcher.js

```diff
@@ -0,0 +1,211 @@
+
+const CHROME_PATH = path.resolve(__dirname, './headless-chromium')
+
+export default class Launcher {
+  constructor (options = {}) {
+    const {
+      chromePath = CHROME_PATH,
+      chromeFlags = [],
+      startingUrl = 'about:blank',
+      port = 0,
+    } = options
+
+    this.tmpDirandPidFileReady = false
+    this.pollInterval = 500
+    this.pidFile = ''
+    this.startingUrl = 'about:blank'
+    this.outFile = null
+    this.errFile = null
+    this.chromePath = CHROME_PATH
+    this.chromeFlags = []
+    this.requestedPort = 0
+    this.userDataDir = ''
+    this.port = 9222
+    this.pid = null
+    this.chrome = undefined
+
+    this.options = options
+    this.startingUrl = startingUrl
+    this.chromeFlags = chromeFlags
+    this.chromePath = chromePath
+    this.requestedPort = port
+  }
+
+  get flags () {
+    return [
+      ...DEFAULT_CHROME_FLAGS,
+      `--remote-debugging-port=${this.port}`,
+      `--user-data-dir=${this.userDataDir}`,
+      '--disable-setuid-sandbox',
+      ...this.chromeFlags,
+      this.startingUrl,
+    ]
+  }
+
+  prepare () {
+    this.userDataDir = this.options.userDataDir || makeTempDir()
+    this.outFile = fs.openSync(`${this.userDataDir}/chrome-out.log`, 'a')
+    this.errFile = fs.openSync(`${this.userDataDir}/chrome-err.log`, 'a')
+    this.pidFile = `${this.userDataDir}/chrome.pid`
+    this.tmpDirandPidFileReady = true
+  }
+
+  // resolves if ready, rejects otherwise
+  isReady () {
+    return new Promise((resolve, reject) => {
+      const client = net.createConnection(this.port)
+
+      client.once('error', (error) => {
+        clearConnection(client)
+        reject(error)
+      })
+
+      client.once('connect', () => {
+        clearConnection(client)
+        resolve()
+      })
+    })
+  }
+
+  // resolves when debugger is ready, rejects after 10 polls
+  waitUntilReady () {
+    const launcher = this
+
+    return new Promise((resolve, reject) => {
+      let retries = 0;
+      (function poll () {
+        debug('Waiting for Chrome', retries)
+
+        launcher
+          .isReady()
+          .then(() => {
+            debug('Started Chrome')
+            resolve()
+          })
+          .catch((error) => {
+            retries += 1
+
+            if (retries > 10) {
+              return reject(error)
+            }
+
+            return delay(launcher.pollInterval).then(poll)
+          })
+      }())
+    })
+  }
+
+  async spawn () {
+    const spawnPromise = new Promise(async (resolve) => {
+      if (this.chrome) {
+        debug(`Chrome already running with pid ${this.chrome.pid}.`)
+        return resolve(this.chrome.pid)
+      }
+
+      const chrome = spawn(this.chromePath, this.flags, {
+        detached: true,
+        stdio: ['ignore', this.outFile, this.errFile],
+      })
+
+      this.chrome = chrome
+
+      // unref the chrome instance, otherwise the lambda process won't end correctly
+      if (chrome.chrome) {
```

Comment:
I just found this block during typescript migration, isn't `chrome.chrome ` always `undefined`?

Reply:
Hm.. That's a good point. It may be a typo and was meant to be `this.chrome`?Given that as currently written it's always `undefined`, maybe the block of code is completely useless and could just be deleted....

Reply:
thanks for checking, and yeah, 👍 for `detached: true` option. 
but according to child process document, we should discuss about this:

event loop will wait for detached process by default even `detached: true` option supplied.
so lambda runtime will not ended. to accomplish ending lambda function without waiting for chrome process exists, we should call `chrome.unref()` even `detached: true` option supplied, to not include the child in its reference count from event loop

From Node.js Child Process Doc:

> By default, the parent will wait for the detached child to exit. To prevent the parent from waiting for a given subprocess, use the subprocess.unref() method. Doing so will cause the parent's event loop to not include the child in its reference count, allowing the parent to exit independently of the child, unless there is an established IPC channel between the child and parent.

Reply:
and it looks that `src/index.js` has same block too. maybe duplicated?

Reply:
Ah. That's probably why it still worked! In that case it could probably safely be removed from `launcher.js`.